### ISSUE TITLE
Knight moves

### ISSUE THREAD

**Issue by wavesinaroom** [wavesinaroom/my_odin_project_knights_travails#2]
Treat all possible moves the knight could make as children in a tree. Don't allow any moves to go off the board.

**Comment by wavesinaroom** [comment]
Array are a lot easier to iterate over than creating an tree as a class. However, I saved the tree as a class version in another branch just in case

**Comment by wavesinaroom** [comment]
I'm gonna spend some time in fixing the inverted coordinates issue in Board class tiles attribute. That bug is causing weird placement on the board that bothers me. I can't get exactly why knight moves are out of the right board boundaries and why I've got an exception when exceding board bottom boundary

**Comment by wavesinaroom** [comment]
Fixed the issue with setting the tiles coordinates. Now, I'm still deciding whether zero based numbering is the best option to go for or not

**Comment by wavesinaroom** [comment]
Changing zero based numbering is nonsense, I just needed to tweak isOutOfBoundaries method if statements to sort out that issue. My focus at this point of the project was changed to testing thouroughly knights possible moves on the board. That was really important to have just the right positions inside the moves array where a recursive method will iterate

**Comment by wavesinaroom** [comment]
Base case for the recursive function is a board full of possible moves 'a' token... well that's my guess so far

**Comment by wavesinaroom** [comment]
A path ends when a knight move goes off the board as the task description says. That means that the get moves recursive calls stop when a knight's move goes off the board. Sounds obvious but I needed to understand that well.

**Comment by wavesinaroom** [comment]
Now that I think of it, getMoves recursion calls also stop when the find a target tile. Howeve, I'd rather focus on getting all the moves first that don't go off the board

**Comment by wavesinaroom** [comment]
Now that I think of it, getMoves recursion calls also stop when the find a target tile. Howeve, I'd rather focus on getting all the moves first that don't go off the board

**Comment by wavesinaroom** [comment]
As long as there's a knight on the board, there'll be moves. That means that getMoves() recursive calls can't be stopped by the moves array length which will never be 0 as a thought previously

**Comment by wavesinaroom** [comment]
Recursion gets stuck at one point. I'm trying to find out why it stays there, something wrong must be going on with its base case

**Comment by wavesinaroom** [comment]
Relying on getting the target as base case creates pits where our knight will never reach it. Recursion calls have to be handled in another way

**Comment by wavesinaroom** [comment]
I've already migrated the array solution to a tree solution hopin that it'd the issue with the endless recursion pits but it did not. I'm gonna take some time to find why the getMoves function recursive calls get stuck at some points

**Comment by wavesinaroom** [comment]
The function is being called even when the target is being found. That means that, if target is found, the algorithm seems to continue from that position onwards creating an infinite loop

**Comment by wavesinaroom** [comment]
After logging the output of each recursive call I found that propably the code generates a tree from the starting knight position so it might be creating an infinite loop. Better be sure by removing the visited position like an algorithm for finding a path to the exit in a maze

**Comment by wavesinaroom** [comment]
I set up a debugger to track getMoves recursive calls

**Comment by wavesinaroom** [comment]
Debbuger steps on each line of code and I'd like to step into each recursive call

**Comment by wavesinaroom** [comment]
Trying to make a tree of knights possible moves doesn't seem to be a good strategy. It easily takes you to infinite loop pits and its to expensive in terms of memory usage. I'd definitely like to quit this approach and try a graph instead which is suggested in stackoverflow

**Comment by wavesinaroom** [comment]
getMoves will alway be an infinite loops as it is right now. However, I can stop it BFS according to a stackoverflow comment

**Comment by wavesinaroom** [comment]
Execution should stop when target is found for the first time

**Comment by wavesinaroom** [comment]
I've just realized that I reached an area out of the scope of this issue. Basically the tree is already, so I have to move on with issue 3 cause the program works like this: 1. Get all knight possible moves from the starting position 2. Run BFS to check if any move reaches target 3. If target's reached stop else create possible moves tree for each current move recursively. 4. Rinse and repeat until achieving goal
